ui/app/controllers/settings/tokens.js

```diff
@@ -66,4 +78,74 @@ export default class Tokens extends Controller {
       }
     );
   }
+
+  // Generate a 20-char nonce, using window.crypto to
+  // create a sufficiently-large output then trimming
+  generateNonce() {
+    let randomArray = new Uint32Array(10);
+    crypto.getRandomValues(randomArray);
+    return randomArray.join('').slice(0, 20);
+  }
+
+  @action redirectToSSO(method) {
+    const provider = method.name;
+    const nonce = this.generateNonce();
+
+    window.localStorage.setItem('nomadOIDCNonce', nonce);
+    window.localStorage.setItem('nomadOIDCAuthMethod', provider);
+
+    method
+      .getAuthURL({
+        AuthMethod: provider,
+        ClientNonce: nonce,
+        RedirectUri: Ember.testing
+          ? this.router.currentURL
+          : window.location.toString(),
+      })
+      .then(({ AuthURL }) => {
+        if (Ember.testing) {
+          this.router.transitionTo(AuthURL.split('/ui')[1]);
+        } else {
+          window.location = AuthURL;
+        }
+      });
+  }
+
+  @tracked code = null;
+  @tracked state = null;
+
+  get isValidatingToken() {
+    if (this.code && this.state === 'success') {
+      this.validateSSO();
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  validateSSO() {
+    this.token
+      .authorizedRequest('/v1/acl/oidc/complete-auth', {
+        method: 'POST',
+        body: JSON.stringify({
+          AuthMethod: window.localStorage.getItem('nomadOIDCAuthMethod'),
+          ClientNonce: window.localStorage.getItem('nomadOIDCNonce'),
+          Code: this.code,
+          State: this.state,
+        }),
+      })
+      .then(async (response) => {
+        if (response.ok) {
+          let json = await response.json();
+          this.token.set('secret', json.ACLToken);
+          this.verifyToken();
+          this.code = null;
+          this.state = null;
+        }
+      });
```

Comment:
`question`: Because we're using `.then` instead of `async/await` this callback will be fired once the request resolves and the rest of the thread of execution will resume? However, we have an `isValidatingToken` that will cause a conditional part of the template to render. I wonder what would happen if we simulate a slow network request. Could we trigger an unhandled error as a result? Would this be a good argument to advocate for `async/await` instead?

Reply:
Great minds think alike! A slow network request is exactly what we're simulating here: https://github.com/hashicorp/nomad/blob/60f7b6073d70ed2e07a105ebb56025bf48aa5eee/ui/mirage/config.js#L955

(timing: 1000 will wait a second before returning our token)

As for .thenning vs async/awaiting, I will make that switch! good suggestion.